### PR TITLE
TM anomaly

### LINKED ISSUE #347: Anomaly removal, cleanup
I have revisited the Anomaly class and I'd like to propose a removal, cleanup of said class, keeping only the `computeRawAnomalyScore()` method, and possible a (new) 
`Anomaly::compute(SDR)` which has only 1 parameter, and handles the bookkeeping of previous pattern for anomaly. 

This will make the code simple, help some confusion in documantation of that class, ... 

- [ ] implement `TM::anomalyScore()`
  - [ ] remove class Anomaly
    - [ ] what happens with AnomalyLikelihood? 
- [ ] move `VectorHelpers::columnsFromCells()` to TM, as the method is only correct for TM's mini-columns, see https://github.com/htm-community/nupic.cpp/pull/331#discussion_r272256537 
  - [ ] implement output `const SDR TM::active/predicted/both`
- [ ] figure predicted vs. pred + act for previous values

### PR DESCRIPTION
**Pull request by breznak**
Anomaly is computed in TM directly, not in Anomaly class

TODO:
- [x] finish implementation :) 
- [x] move AnomalyTests to TM.anomalyTests
  - [ ] new AnomalyLikelihood tests -> must be fixed in separate PR.
- [x] remove Anomaly class
  - [x] remove computeRawAnomalyScore() vector version, keep only SDR
- [x] bindings for TM::getAnomalyScore()
- [x] serialization
- [x] doc
- [ ] ~~waiting for #442 to resolve issue with extraActive, causing SDR to fail. Triggered by "testExtraActive " test~~

Fixes #347

### CODE REVIEW

File: src/nupic/algorithms/TemporalMemory.cpp [hunk 1]
```diff
@@ -646,6 +647,12 @@ void TemporalMemory::compute(const SDR &activeColumns, bool learn,
 void TemporalMemory::compute(const SDR &activeColumns, bool learn) {
   activateDendrites(learn);
   activateCells(activeColumns, learn);
+  //anomaly computation
+  anomaly_.currentActiveColumns = this->cellsToColumns(activeCells_); //FIXME cellsToColumns from PR#397
```
Comment (breznak) [inline src/nupic/algorithms/TemporalMemory.cpp]: waiting for #397 for cellsToColumns

File: src/nupic/algorithms/TemporalMemory.cpp [hunk 2]
```diff
@@ -646,6 +647,12 @@ void TemporalMemory::compute(const SDR &activeColumns, bool learn,
 void TemporalMemory::compute(const SDR &activeColumns, bool learn) {
   activateDendrites(learn);
   activateCells(activeColumns, learn);
+  //anomaly computation
+  anomaly_.currentActiveColumns = this->cellsToColumns(activeCells_); //FIXME cellsToColumns from PR#397 
+  anomaly_.previouslyPredictedColumns = this->cellsToColumns(getPredictiveCells()); //TODO move to T-1
+  anomaly_.score = nupic::algorithms::anomaly::computeRawAnomalyScore(
```
Comment (breznak) [inline src/nupic/algorithms/TemporalMemory.cpp]: should I move the method to TM too? I'm inclining to keeping it as a static method in Anomaly.hpp

File: src/nupic/algorithms/TemporalMemory.cpp [hunk 3]
```diff
@@ -646,6 +647,12 @@ void TemporalMemory::compute(const SDR &activeColumns, bool learn,
 void TemporalMemory::compute(const SDR &activeColumns, bool learn) {
```
Comment (breznak) [inline src/nupic/algorithms/TemporalMemory.cpp]: @ctrl-z-9000-times I'd also prefer the TM.anomalyScore (const float&), to the TM::getAnomalyScore()

the problem is that with the former, I'll work if the user calls compute() but not if they call activateCells/Dendrites separately. Is this worth, or just use getAnomalyScore and implement a check with warning there?

File: src/nupic/algorithms/TemporalMemory.cpp (hunk 2 above)
Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/TemporalMemory.cpp]: > I'm inclining to keeping it as a static method in Anomaly.hpp

That's fine.  It's easier to test if its separate too.

File: src/nupic/algorithms/TemporalMemory.cpp (hunk 3 above)
Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/TemporalMemory.cpp]: Would this issue go away if you put the anomaly code inside of `activateCells`?

**Comment by breznak** [comment]
Requires #437 for fixes, but the core idea can be reviewed here already, when you have time. Thanks

**Comment by breznak** [comment]
> Requires #437 for fixes, but the core idea can be reviewed here already, when you have time. Thanks 

Still waiting for #442 to resolve the issue. SDR failing to create when activateDendrites is used with extraActive; TM.anomaly uses SDR in its update() method.

**Comment by breznak** [comment]
Update: I came with a workaround for 

> Still waiting for #442 to resolve the issue. SDR failing to create when activateDendrites is used with extraActive 

now, TM does not support anomaly for cases when extra inputs (param extra>0) are used. Otherwise anomaly runs just fine.

File: src/nupic/algorithms/Anomaly.cpp [hunk 4]
```diff
@@ -42,92 +35,18 @@ namespace anomaly {
...
-
   // Calculate and return percent of active columns that were not predicted.
   SDR both(active.dimensions);
   both.intersection(active, predicted);
 
   return (active.getSum() - both.getSum()) / Real(active.getSum());
 }
 
-Real computeRawAnomalyScore(vector<UInt>& active,
-                            vector<UInt>& predicted)
```
Comment (ctrl-z-9000-times) [inline src/nupic/algorithms/Anomaly.cpp]: Can we keep this overload?

**Comment by ctrl-z-9000-times** [comment]
I would also like to use the anomaly likelyhood code in the TM class.  The original publication of NAB (https://arxiv.org/pdf/1510.03336.pdf) uses the likelihood code.

We could create a new API for this, like:
```C++
class AbstractAnomaly{
    virtual Real compute(const sdr::SDR& active, 
                         const sdr::SDR& predicted) = nullptr;
};
```
Or
```C++
typedef std::function<Real (const SDR&, const SDR&)> AbstractAnomaly;
```
And then let the user specify the function (Anomaly or AnomalyLikelihood) and its parameters.  

Alternatively, if one of the three methods (Anomaly, AnomalyLikelihood, Weighted) is objectively the best then we should bake it into the TM, and not worry the users about it.

File: src/nupic/algorithms/Anomaly.cpp (hunk 4 above)
Comment (breznak) [reply src/nupic/algorithms/Anomaly.cpp]: > This is where I keep my notes about how the SDR class should be doing set intersections.

We could keep it, 
or keep the code just as a commet, if you're not actively developing on that? 
And ideally move the comment to SDR.intersection() ?

File: src/nupic/regions/TMRegion.cpp [hunk 5]
```diff
@@ -248,12 +239,9 @@ void TMRegion::compute() {
...
     SDR& sdr = out->getData().getSDR();
-    if (args_.orColumnOutputs) { //aggregate to columns
-      tm_->getActiveCells(sdr);
-      SDR cols = tm_->cellsToColumns(sdr);
-      sdr.setSparse(cols.getSparse());
-    } else { //output as cells
-      tm_->getActiveCells(sdr);
+    tm_->getActiveCells(sdr); //active cells
+    if (args_.orColumnOutputs) { //output as columns
+      sdr = tm_->cellsToColumns(sdr);
```
Comment (dkeeney) [inline src/nupic/regions/TMRegion.cpp]: Should we add another output to TMRegion to provide access to anomaly scores, etc?

File: src/nupic/algorithms/Anomaly.cpp (hunk 4 above)
Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/Anomaly.cpp]: Ok, I we can drop this method, but the rest of my criticism stands.
- Friend classes are confusing and usually not necessary.
- This doesn't include anomaly likelihood.
- External predictive inputs don't work?  I read through the code and it looks like they should just work, if you remove the checks that disable it.

File: src/nupic/regions/TMRegion.cpp (hunk 5 above)
Comment (breznak) [reply src/nupic/regions/TMRegion.cpp]: We can decide this later when discussing "anomaly region", but yes, anomaly score would be moved to TM class, so it'll end up as a new part of existing TMRegion.

File: src/nupic/algorithms/Anomaly.cpp (hunk 4 above)
Comment (breznak) [reply src/nupic/algorithms/Anomaly.cpp]: > Instead of removing the anomaly class and adding a different class, just reuse the existing anomaly class. 

I thought you wanted me to get rid of the old Anomaly class and replace it with as simple as possibleimplementation in TM (?), which I ended up liking. 

1. I'm OK to revive the old Anomaly class, with a few changes *). 
2. Or I can just move the code TMAnomaly to Anomaly.hpp and be done with it. 
Which of those would you prefer? 

> Friend classes are confusing and usually not necessary. 

I'll avoid 'friend access' when moving to a separate new/old file. 

> This doesn't include anomaly likelihood. 

Yes, and it is intended. And about other refactoring in Anomaly class *): 
- I've implemented Anomaly class in the past to simplify various options with anomaly (thresholds, likelihood, moving average,...) 
- now with later usage experience, I don;t think it's a good design, and as those features can be implemented as "one liners" with current code, I don't think it should be provided with anomaly  class. "Let TM.getAnomalyScore() be the simple raw anomaly" (#A), just conveniently computed. And let the user do what they need later with it. 
  - `thereshold`: simple `if(score > XXX) return 1 else return 0;`
  - `likelihood`: 
    - (probably even does not truly work) 
    - in Anomaly had a FIXME note that it's not truly configurable. And should be separated. 
    - I could use the approach with abstract class `AbstractAnomaly::compute(SDR, SDR)`, but it's complicating things, see argument #A
    - AnomalyLikelihood is not even a true anomaly (computed from 2 SDRs), but a likelihood of anomalies, thus computed AFTER and from anomaly score. So `score = TM.getAnomalyScore(); Likelihood::anomalyProbability(score, timestep);` 
  - `moving average`: again, simple `MovingAverage ma(5); ma.add(score); ma.get();`

TL;DR: the proper configuration and ordering is too complicated in a wrapper Anomaly class, while relatively easy hand-tailored by user (with tools we provide). 

> External predictive inputs don't work? I read through the code and it looks like they should just work, if you remove the checks that disable it.

Externals break it in the way the they add "cell indices out of scope of the current TM" (idx in <numberOfCells() , numOfCells + extra>). 
See for yourself in SDR::setSparseInplace(), you must use Debug build to trigger, that's why it's failing only on the OSX CI. I'm proposing to proceed as is, and discuss the fixes on external inputs in a separate issue or #442

File: src/nupic/regions/TMRegion.cpp (hunk 5 above)
Comment (breznak) [reply src/nupic/regions/TMRegion.cpp]: @dkeeney if you have time, would you mind pushing a commit that adds a field "float anomalyScore" to TMRegion, and is computed from `TM.getAnomalyScore()`, please? 
I could do it but Regions are still not my cup of coffee.

Comment (dkeeney) [reply src/nupic/regions/TMRegion.cpp]: Understand.  Yes, I can do this.
But rather than a field I suggest that it be a new output. That way, the Anomaly score can be passed to some other module after each iteration.  It would be cool if we had a plot region to send it to.

If we just make it a field that can be queried then the user needs to add a callback to get it at each iteration.  We can provide that as well but it would not be as useful.

I will take a look at the old .py code to see what the old python Anomaly Region did.  I assume there was one.

Comment (breznak) [reply src/nupic/regions/TMRegion.cpp]: Thank you! 

> But rather than a field I suggest that it be a new output. That way, the Anomaly score can be passed to some other module after each iteration. 

Ok, this is the better solution. Would it be a problem that the output is different? (1 integer, instead of TM::getColumnDimensions(), or TM::getColumnDims + cellsPerColumn? 

> I will take a look at the old .py code to see what the old python Anomaly Region did. I assume there was one.

there wasn't any AnomalyRegion

File: src/nupic/algorithms/Anomaly.cpp (hunk 4 above)
Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/Anomaly.cpp]: > What would you prefer?

My goal is to have convenient way to print out everything interesting about the TM to a human readable string, and from it to determine if the TM is working.  I see the anomaly metric as a diagnostic tool.
 * I want to know the mean & standard deviation of the anomaly.  The anomaly likelihood also uses this info.
 * It should be enabled by default. 

> The proper configuration and ordering is too complicated.

Agreed,
 * The user should be responsible for discarding anomaly readings when they haven't finished training the network.
 * The user should be responsible for applying the threshold.
* AnomalyLikelihood needs 1 parameter to do what it does, and it has 4 or 5.

**Comment by breznak** [comment]
Thank you, David! I'm quite happy with the `TM.anomaly` as it is now :smile: 

Two remaining issues: 
- [ ] anomaly output for TMRegion
- [ ] anomaly works with TM::compute() but breaks with activateCells. 
  - I'd prefer moving activateCells to protected.
  - or we could track TM.anomaly with some iteration number, and NTA_THROW if mismatched

**Comment by dkeeney** [comment]
> adds a field "float anomalyScore" to TMRegion, and is computed from TM.getAnomalyScore()

I assume you still want this new function.
But we did not decide if you also want anomaly to be a Region output.

**Comment by breznak** [comment]
yes, but now it's called just `TM.anomaly` 

> But we did not decide if you also want anomaly to be a Region output. 

I agree with your suggestion that output would be a better choice.

**Comment by dkeeney** [comment]
Ok, I am creating a new PR to do this.

**Comment by dkeeney** [comment]
For Outputs we currently have:
- bottomUpOut  which is from TM::getActiveCells()  which is an SDR of either numberOfCols if orColumnOutputs is set or number of cells by default.
- activeCells   (which is always an SDR of number of cells)
- predictedActiveCells which are the Winners
- anomaly  (which is the new output)

With all of the recent changes, is this still the expected outputs from TMRegion.

**Comment by breznak** [comment]
> With all of the recent changes, is this still the expected outputs from TMRegion. 

yes. 
- anomaly would be 1x1 scalar

> predictedActiveCells which are the Winners 

might change name to `winnerCells` are winners

- might add `predictiveCells` populated with tm.getPredictiveCells()

- probably all of the `*Cells` should use the orColumnOutputs logic (rename that to asColumns)

**Comment by ctrl-z-9000-times** [comment]
> predictedActiveCells which are the Winners

This is not 100% correct.  When a mini-column bursts because it was unpredicted (this is an anomaly!), then a winner cell is arbitrarily chosen.

**Comment by dkeeney** [comment]
Changing the name of predictedActiveCells will break the API.  But we are changing so many other things, now is probably a good time to change this as well.

I was going to get this out in just a few min but but with these name changes this will take longer. My wife says I need to go wash the windows...company coming next week.   I will try to finish this PR later today.

**Comment by breznak** [comment]
> Changing the name of predictedActiveCells will break the API. 

hmm, we wanted to keep NetworkAPI as stable as possible. 

I'm not sure what @ctrl-z-9000-times meant with the comment 
> This is not 100% correct. When a mini-column bursts because it was unpredicted (this is an anomaly!), then a winner cell is arbitrarily chosen. 

if it's just the comment about name vs winner cells, we should keep the old output name, and add this note to the description.

**Comment by dkeeney** [comment]
Ok, windows are washed.  On to more interesting things...
- Lets see, a second output called `predictiveCells`
- orColumnOutputs, lets not change the field name
- PredictedActiveCells will remain that name.  Comments already say its the Winners.

**Comment by dkeeney** [comment]
What comment should I add to the `predictiveCells` output?  Is "The cells that were predicted" or is this "The cells that predict the next set of active cells".

Also, The vector overload of activateDendrites( ) does not appear to be implemented.
```
  void activateDendrites(const bool learn = true,
                         const vector<UInt> &extraActive  = {std::numeric_limits<UInt>::max()},
                         const vector<UInt> &extraWinners = {std::numeric_limits<UInt>::max()});
```

**Comment by breznak** [comment]
"Cells predicted to be active in the next step" ?

> Also, The vector overload of activateDendrites( ) does not appear to be implemented. 

yes, there's only SDR version now. activateDendrites(bool), or activateDendrites(bool, SDR, SDR)

**Comment by dkeeney** [comment]
oh, just needed to pull from master.
